core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -328,35 +311,42 @@ public boolean remove(Object object) {
    * @param object the object to be removed
    */
   private void removeFromIndices(T object) {
-    for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> fieldInfo :
-        mIndexMap.entrySet()) {
-      Object fieldValue = fieldInfo.getKey().getFieldValue(object);
-      ConcurrentHashMap<Object, ConcurrentHashSet<T>> index = fieldInfo.getValue();
-      ConcurrentHashSet<T> objSet = index.get(fieldValue);
-      if (objSet != null) {
-        objSet.remove(object);
-      }
+    for (Map.Entry<String, FieldIndex<T>> fieldInfo : mIndices.entrySet()) {
+      fieldInfo.getValue().remove(object);
     }
   }
 
   /**
-   * Removes the subset of objects with the specified field value.
+   * Removes the object with the specified unique index field value.
    *
-   * @param index the field index
+   * @param indexName the field index
    * @param value the field value
    * @return the number of objects removed
    */
-  public int removeByField(FieldIndex<T> index, Object value) {
-    ConcurrentHashSet<T> toRemove = getByFieldInternal(index, value);
-    if (toRemove == null) {
-      return 0;
-    }
+  public int removeByField(String indexName, Object value) {
     int removed = 0;
-    for (T o : toRemove) {
-      if (remove(o)) {
+    FieldIndex<T> index = mIndices.get(indexName);
```

Comment:
yeah, I think just getting the `set` of values (regardless of unique index or not) and deleting them should work.